File: src/main/java/org/distril/beengine/util/ChunkUtils.java
package org.distril.beengine.util;


import com.google.common.base.Preconditions;

public class ChunkUtils {

	public static void checkBounds(int x, int y, int z) {
		Preconditions.checkElementIndex(x, 16, "X coordinate");
		Preconditions.checkElementIndex(y, 256, "Y coordinate");
		Preconditions.checkElementIndex(z, 16, "Z coordinate");
	}

	public static int fromKeyX(long key) {
		return (int) (key >> 32);
	}

	public static int fromKeyZ(long key) {
		return (int) key;
	}

	public static long key(int x, int z) {
		return (((long) x) << 32) | (z & 0xffffffffL);
	}
}

File: src/main/kotlin/org/distril/beengine/util/ChunkUtils.kt
package org.distril.beengine.util

import org.distril.beengine.util.Utils.requireInRange

object ChunkUtils {

	fun checkBounds(x: Int, y: Int, z: Int) {
		requireInRange(x, maxValue = 16, name = "X")
		requireInRange(y, maxValue = 256, name = "Y")
		requireInRange(z, maxValue = 16, name = "Z")
	}

	fun decodeX(chunkKey: Long) = (chunkKey shr 32).toInt()

	fun decodeZ(chunkKey: Long) = chunkKey.toInt()

	fun encode(chunkX: Int, chunkZ: Int) = chunkX.toLong() shl 32 or (chunkZ.toLong() and 0xffffffffL)
}
